service/cloud/gcp_streamer.go

```diff
@@ -18,129 +18,149 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"sort"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"cloud.google.com/go/storage"
 	"github.com/fxamacker/cbor/v2"
 	"github.com/gammazero/deque"
-	"github.com/go-playground/validator/v10"
 	"github.com/rs/zerolog"
-	"google.golang.org/api/iterator"
 
 	"github.com/onflow/flow-go/engine/execution/computation/computer/uploader"
+	"github.com/onflow/flow-go/model/flow"
 
 	"github.com/optakt/flow-dps/models/dps"
 )
 
 type GCPStreamer struct {
-	log      zerolog.Logger
-	bucket   *storage.BucketHandle
-	validate *validator.Validate
-	buffer   *deque.Deque
-	last     time.Time
-	wg       *sync.WaitGroup
-	limit    uint
+	log    zerolog.Logger
+	ext    string
+	bucket *storage.BucketHandle
+	queue  *deque.Deque
+	buffer *deque.Deque
+	last   time.Time
+	busy   uint32
+	wg     *sync.WaitGroup
+	limit  uint
 }
 
 func NewGCPStreamer(log zerolog.Logger, bucket *storage.BucketHandle) *GCPStreamer {
 
 	g := GCPStreamer{
-		log:      log,
-		bucket:   bucket,
-		validate: validator.New(),
-		buffer:   deque.New(),
-		last:     time.Time{},
-		wg:       &sync.WaitGroup{},
-		limit:    8,
+		log:    log.With().Str("component", "gcp_streamer").Logger(),
+		ext:    ".cbor",
+		bucket: bucket,
+		queue:  deque.New(),
+		buffer: deque.New(),
+		last:   time.Time{},
+		busy:   0,
```

Comment:
The point is that it's _explicit_ :grin:

```
Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
```